crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2660,6 +2660,55 @@ impl<'db> TypeInferenceBuilder<'db> {
                 self.infer_binary_type_comparison(left, op, KnownClass::Str.to_instance(self.db))
             }
 
+            (Type::Tuple(lhs), Type::Tuple(rhs)) => {
+                // Note: This only works on heterogeneous tuple types.
+                let lhs_elements = lhs.elements(self.db).as_ref();
+                let rhs_elements = rhs.elements(self.db).as_ref();
+
+                match op {
+                    ast::CmpOp::Eq
+                    | ast::CmpOp::NotEq
+                    | ast::CmpOp::Lt
+                    | ast::CmpOp::LtE
+                    | ast::CmpOp::Gt
+                    | ast::CmpOp::GtE => {
+                        self.infer_lexicographic_type_comparison(lhs_elements, op, rhs_elements)
+                    }
+                    ast::CmpOp::In | ast::CmpOp::NotIn => {
+                        // - `[ast::CmpOp::In]`: returns `true` if the element is definitely in the tuple, otherwise `bool`
+                        // - `[ast::CmpOp::NotIn]`: returns `false` if the element is definitely in the tuple, otherwise `bool`
+                        if rhs_elements.iter().any(|ty| {
+                            let eq_result_ty = self.infer_binary_type_comparison(
+                                Type::Tuple(lhs),
+                                ast::CmpOp::Eq,
+                                *ty,
+                            );
+
+                            matches!(eq_result_ty, Some(Type::BooleanLiteral(true)))
+                        }) {
+                            Some(Type::BooleanLiteral(op.is_in()))
+                        } else {
+                            Some(KnownClass::Bool.to_instance(self.db))
+                        }
+                    }
+                    ast::CmpOp::Is | ast::CmpOp::IsNot => {
+                        // - `[ast::CmpOp::Is]`: returns `false` if the elements are definitely unequal, otherwise `bool`
+                        // - `[ast::CmpOp::IsNot]`: returns `true` if the elements are definitely unequal, otherwise `bool`
+                        let eq_result_ty = self.infer_lexicographic_type_comparison(
+                            lhs_elements,
+                            ast::CmpOp::Eq,
+                            rhs_elements,
+                        );
+
+                        if let Some(Type::BooleanLiteral(false)) = eq_result_ty {
+                            Some(Type::BooleanLiteral(!op.is_is()))
+                        } else {
+                            Some(KnownClass::Bool.to_instance(self.db))
+                        }
```

Comment:
I’m unsure if implementing this part in such a complex way is the best approach. Couldn’t we simply use `Some(KnownClass::Bool.to_instance(self.db))` instead? While the current implementation can provide powerful inference in limited specific cases, the cost of maintenance might outweigh the benefits. Feedback on this part would be appreciated.

Reply:
This is a great question. My feeling is that it is worth it. I think once we get the logic correct the maintenance cost will not be high. And there will be a few common cases where this will definitely make a difference. For instance, we will need to implement detailed comparison of `sys.version_info` tuples anyway; I think it's much nicer if this just uses general support for tuple comparison inference rather than being handled as a special case.

Reply:
Thanks for the insight! I hadn’t thought about `sys.version_info`.
I’m excited to see how powerful the type inference will become 👍